### PR TITLE
Make heroes fly on pushReplacement

### LINKED ISSUE #28041: Hero animation not triggered with Navigator.pushReplacement
Hero animations are not triggered when replacing a route. I'd like to use that for example at the end of an on-boarding flow where the 'main' page replaces the `on-boarding` page.

~It can be easily fixed by implementing `HeroController.didReplace` but it's probably good to discuss here whether it is desired.~

Minimal example:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(
        home: Page1(),
      );
}

class Page1 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: GestureDetector(
          onTap: () => Navigator.of(context).pushReplacement(MaterialPageRoute(builder: (context) => Page2())),
          child: Hero(
            tag: "superman",
            child: Container(
              width: 40,
              height: 40,
              color: Colors.red,
            ),
          ),
        ),
      ),
    );
  }
}

class Page2 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Hero(tag: "superman", child: Container(width: 60, height: 60, color: Colors.red)),
      ),
    );
  }
}
```

```
[✓] Flutter (Channel master, v1.2.2-pre.21, on Mac OS X 10.14.3 18D109, locale en-NL)
    • Flutter version 1.2.2-pre.21 at /Users/spkersten/Development/flutter
    • Framework revision 035e0765cc (3 hours ago), 2019-02-16 01:48:28 -0500
    • Engine revision f37b09a11b
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 c92d5ca288)
```

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

Before this change Heroes wouldn't fly when `Navigator.pushReplacement` was used because it triggered `NavigatorObserver.didReplace`, which was unimplemented by the HeroController.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/28041.

## Tests

I added the following tests:

* Added a regression test for https://github.com/flutter/flutter/issues/28041.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 1]
```diff
@@ -618,6 +618,15 @@ class HeroController extends NavigatorObserver {
     _maybeStartHeroTransition(route, previousRoute, HeroFlightDirection.pop, false);
   }
 
+  @override
+  void didReplace({ Route<dynamic> newRoute, Route<dynamic> oldRoute }) {
+    assert(navigator != null);
+    if (newRoute?.isCurrent == true) {
+      // Only run hero animations if top-most route got replaced.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart]: if => if the

**Review by HansMuller — APPROVED** [review]
LGTM. Feel free to land this without the trivial copy-edit I suggested.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done

**Comment by Hixie** [comment]
This (meaning Navigator and especially NavigatorObserver) really is the worst API.

**Pull request merged**